src/box/lua/config/applier/credentials.lua

```diff
@@ -1,30 +1,326 @@
+   objects of a type.
+
+privs:
+ - read
+ - write
+ - execute
+   - lua_eval
+   - lua_call
+   - sql
+ - session
+ - usage
+ - create
+ - drop
+ - alter
+ - reference
+ - trigger
+ - insert
+ - update
+ - delete
+
+Examples:
+- - box.schema.user.grant('myuser', 'execute', 'function', 'myfunc')
+  - ['function']['myfunc']['execute'] = true
+  - grant execute of myfunc
+
+- - box.schema.user.grant('myuser', 'execute', 'function')
+  - ['function']['']['execute'] = true
+  - grant execute of all registered functions
+
+- - box.schema.user.grant('myuser', 'read', 'universe')
+  - ['universe']['']['read'] = true
+  - grant read to universe
+
+- - box.schema.user.grant('myuser', 'execute', 'role', 'super')
+  - ['role']['super']['execute'] = true
+  - equivalent to granting a role to myuser
+
+]]--
+
+local function privileges_box_convert(privileges)
+    privileges = privileges or {}
+    assert(type(privileges) == 'table')
+
+    local res = {
+        ['user'] = {},
+        ['role'] = {},
+        ['space'] = {},
+        ['function'] = {},
+        ['sequence'] = {},
+        ['universe'] = {},
+    }
+
+    for _, priv in ipairs(privileges) do
+        local perms, obj_type, obj_name = unpack(priv)
+        obj_name = obj_name or ''
+
+        res[obj_type][obj_name] = res[obj_type][obj_name] or {}
+
+        for _, perm in ipairs(perms:split(',')) do
+            res[obj_type][obj_name][perm] = true
+        end
+    end
+
+    return res
+end
+
+local function privileges_config_convert(config_data)
+    local privileges = config_data.privileges or {}
+    assert(type(privileges) == 'table')
+
+    local intermediate = {
+        ['user'] = {},
+        ['role'] = {},
+        ['space'] = {},
+        ['function'] = {},
+        ['sequence'] = {},
+        ['universe'] = {},
+    }
+
+    local function add_permission(intermediate, perm, obj_def, obj_type)
+        obj_def = obj_def or {}
+        -- XXX maybe rename 'all' to something better
+        if obj_def[1] == 'all' then
+            intermediate[obj_type][''] = intermediate[obj_type][''] or {}
+            intermediate[obj_type][''][perm] = true
+            return
+        end
+
+        for _, obj_name  in ipairs(obj_def) do
+            intermediate[obj_type][obj_name] =
+                intermediate[obj_type][obj_name] or {}
+            intermediate[obj_type][obj_name][perm] = true
+        end
+    end
+
+    for _, priv in ipairs(privileges) do
+        for _, perm in ipairs(priv.permissions) do
+            if priv.universe then
+                intermediate['universe'][''] =
+                        intermediate['universe'][''] or {}
+                intermediate['universe'][''][perm] = true
+            else
+                add_permission(intermediate, perm, priv.spaces, 'space')
+                add_permission(intermediate, perm, priv.functions, 'function')
+                add_permission(intermediate, perm, priv.sequences, 'sequence')
```

Comment:
Can't permissions on `universe` be set together with other privileges? I would handle them all, without guessing that if `universe` is present, we shouldn't process others.

Reply:
Inside schema it doesn't feel very natural to have both `'universe'` and some objects in one privilege. But as far as I can recall, there is no verification and I don't think that applier should just ignore those objects. So yea, lets add them all here.

Reply:
Generally, I think, that the same scenario as you mentioned above can be applied here too:
```yaml
credentials:
  users:
    myuser:
      privileges:
      - permissions: [execute]
        functions:
        - some.func
        # Temporary grant all functions for testing purposes.
        universe: true
```

So, verification can be left as is too.